### PR TITLE
chore(typescript): Upgrade to 2.0 & Types

### PR DESCRIPTION
**Pull request by MarkPieszak**
### ! Work in progress !

Upgraded all Project Templates to TS2.0beta, `@types`, removed all remnants of `tsd`, `typings` folders, etc.
Left Universal specific custom types in `types/custom-types.d.ts` for the ng2 template.

> Note: TypeScript2 is still in `beta`, Visual Studio / VSCode / etc need to be setup to realize they're using a beta/nightly build of TypeScript to take advantage of 2.0 features.

### Tested & Working:

- [x] Angular2Spa
- [x] KnockoutSpa
- [x] ReactSpa
- [x] WebApplicationBasic

## ReactReduxSpa 
it looks like the project needs to be updated to the latest API, `@types` here cause many errors, since its based on an older API.

- [ ]  (*Issues remaining with `certain react @types / Store / Dispatch`*)
   - node_modules/**redux-typed/StrongActions.d.ts** `(16,16): error TS2314: Generic type 'Dispatch<S>' requires 1 type argument(s).` (can be fixed with PR #256 )
   - **@types/react-router-redux/index.d.ts** error TS2305: Module '"history"' has no exported member 'LocationDescriptor'. (possible Type definition issue? [Opened here](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/10686))
   - **./ClientApp/boot-client.tsx** error TS2322: Type 'ReactRouterReduxHistory' is not assignable to type 'History'. Property 'listenBefore' is missing in type 'ReactRouterReduxHistory'.

> ERROR in C:\Users\mark.pieszak\Documents\GitHub\JavaScriptServices\templates\ReactReduxSpa\node_modules\redux-typed\StrongActions.d.ts
> (16,16): error TS2314: Generic type 'Dispatch<S>' requires 1 type argument(s).
> 
> ERROR in C:\Users\mark.pieszak\Documents\GitHub\JavaScriptServices\templates\ReactReduxSpa\types\custom-types.d.ts
> (7,1): error TS2688: Cannot find type definition file for 'react-router/history.d.ts'.
> 
> ERROR in ./ClientApp/configureStore.ts
> (12,39): error TS2346: Supplied parameters do not match any signature of call target.
> 
> ERROR in ./ClientApp/configureStore.ts
> (19,75): error TS2686: Identifier 'Redux' must be imported from a module
> 
> ERROR in ./ClientApp/configureStore.ts
> (19,75): error TS2314: Generic type 'Store<S>' requires 1 type argument(s).
> 
> ERROR in ./ClientApp/configureStore.ts
> (33,91): error TS2686: Identifier 'Redux' must be imported from a module

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
Thanks for this! I'll merge it as soon as TypeScript 2.0 ships. It will be great to get all those `.d.ts` files out of the project 👍

**Comment by MarkPieszak** [comment]
Glad to help 😄 
Sounds like a plan! 
Doing some finishes touches on the PR at the moment. Almost finished testing & fixing all the templates.

**Comment by MarkPieszak** [comment]
Just fixed the merged conflicts, @SteveSandersonMS 
Can we merge this before merging the Angular2 final + Universal updates?

**Comment by stephtr** [comment]
If it helps: _"error TS2686: Identifier 'Redux' must be imported from a module"_ seems to be related with [TypeScript #10638](https://github.com/Microsoft/TypeScript/issues/10638). Using typescript@2.1.0-dev.20160918 these errors vanish.

_"(12,39): error TS2346: Supplied parameters do not match any signature of call target"_ can be fixed by changing windowIfDefined.devToolsExtension to
`windowIfDefined.devToolsExtension as () => (a: any) => any`

**Comment by MarkPieszak** [comment]
Thank you @stephtr I'll take a look! I wasn't sure if we would want to use typescript@next since it cold e very unstable though?

One of the other main problems is that the store API in the one project is very outdated model (based on an old tsd commit).  Hoping I can spend some time with Steve to look through the ReactRedux changes needed.

Other than that, I was able to get all the other projects running without errors luckily.

**Comment by tinchou** [comment]
Maybe you can take a look at @types if you're updating to Typescript 2 :)

**Comment by MarkPieszak** [comment]
Course! Already have them in here 😃

**Comment by stephtr** [comment]
One change made in Redux 3.6 is _Improve typings for `compose` function_, therefore a type definition is needed for devToolsExtension.

Using only [a few changes](https://github.com/MarkPieszak/JavaScriptServices/compare/TS2-Types...stephtr:TS2-Types) ReactReduxSpa runs using TypeScript 2.0 (typescript@next not necessary) and @types.
While experimenting I noticed that TypeScript seems to use the typing included in the redux-package and not from @types/redux.
types/custom-types.d.ts is not needed when including node_modules/@types/webpack/webpack-env.d.ts either via tsconfig.json -> types (not working for me) or by referencing it in node_modules/@types/webpack/index.d.ts (see (DefinitelyTyped #10978)[https://github.com/DefinitelyTyped/DefinitelyTyped/issues/10978]).

**Comment by MarkPieszak** [comment]
Hoping to get back to this one later today so we can get this merged in there since TS2 finally shipped :)
@SteveSandersonMS maybe we can link up later to figure out those ReactRedux issues? Seems the API has changed quite a bit for compose and a few other things.

File: templates/ReactReduxSpa/package.json [hunk 1]
```diff
@@ -1,7 +1,18 @@
 {
-  "name": "WebApplicationBasic",
+  "name": "react-redux-spa",
   "version": "0.0.0",
   "dependencies": {
```
Comment (antmdvs) [inline templates/ReactReduxSpa/package.json]: Not sure why I can't comment on line 40 (it's grayed out), but bump typescript to v2.
`"typescript": "^2.0.3",`

**Comment by stephtr** [comment]
Which issues are still left (except redux-typed)? Because in my fork it is working ([stephtr/JavaScriptServices](https://github.com/stephtr/JavaScriptServices/tree/TS2-Types/templates/ReactReduxSpa)).

File: templates/ReactReduxSpa/package.json (hunk 1 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/package.json]: Great idea! Let me add that in there.

**Comment by MarkPieszak** [comment]
Closed in favor of #362

**Pull request closed without merging**